server/spec/mutations/grid_services/update_spec.rb

```diff
@@ -93,6 +93,111 @@
       ).run
       expect(outcome.success?).to be(true)
     end
+
+    context 'volumes' do
+      context 'stateless service' do
+        it 'allows to add non-named volume' do
+          outcome = described_class.new(
+            grid_service: redis_service,
+            volumes: ['/foo']
+          ).run
+          expect(outcome.success?).to be_truthy
+          expect(outcome.result.volumes).to eq(['/foo'])
+        end
+
+        it 'allows to add named volume' do
+          outcome = described_class.new(
+            grid_service: redis_service,
+            volumes: ['foo:/foo']
+          ).run
+          expect(outcome.success?).to be_truthy
+          expect(outcome.result.volumes).to eq(['foo:/foo'])
+        end
+
+        it 'allows to add bind mounted volume' do
+          outcome = described_class.new(
+            grid_service: redis_service,
+            volumes: ['/foo:/foo']
+          ).run
+          expect(outcome.success?).to be_truthy
+          expect(outcome.result.volumes).to eq(['/foo:/foo'])
+        end
+      end
+
+      context 'stateful service' do
+        let(:stateful_service) do
+          GridService.create(
+            grid: grid, name: 'redis', image_name: 'redis:2.8', stateful: true,
+            volumes: ['/data']
+          )
+        end
+
+        it 'does not allow to add non-named volume' do
+          outcome = described_class.new(
+            grid_service: stateful_service,
+            volumes: ['/data', '/foo']
+          ).run
+          expect(outcome.success?).to be_falsey
+        end
+
+        it 'allows to add named volume' do
+          outcome = described_class.new(
+            grid_service: stateful_service,
+            volumes: ['/data', 'foo:/foo']
+          ).run
+          expect(outcome.success?).to be_truthy
+          expect(outcome.result.volumes).to eq(['/data', 'foo:/foo'])
+        end
+
+        it 'allows to add bind mounted volume' do
+          outcome = described_class.new(
+            grid_service: stateful_service,
+            volumes: ['/data', '/foo:/foo']
+          ).run
+          expect(outcome.success?).to be_truthy
+          expect(outcome.result.volumes).to eq(['/data', '/foo:/foo'])
+        end
+
+        it 'allows to remove a volume' do
```

Comment:
Removing an anonymous `volumes: /var/lib/test` from a stateful service doesn't actually remove volumes present when it was initially deployed, since those anonymous volumes will still be configured on the `*-volumes` container, and will still be visible to the re-deployed service container.

```
core@core-01 ~ $ docker inspect -f '{{.Config.Volumes}}' wordpress.wordpress-1
map[/var/www/html:{}]
core@core-01 ~ $ docker inspect -f '{{range .Mounts}}{{.Destination}} {{end}}' wordpress.wordpress-1-volumes
/var/lib/test /var/www/html 
core@core-01 ~ $ docker exec wordpress.wordpress-1 ls -la /var/lib/test/
total 20
drwxr-xr-x. 2 root root 4096 Feb  1 13:29 .
drwxr-xr-x. 1 root root 4096 Feb  1 13:29 ..
-rw-r--r--. 1 root root    0 Feb  1 13:29 asdf
```

Probably not much of an issue, but that's how it behaves :)